sway/ipc-server.c

```diff
@@ -322,6 +336,266 @@ void ipc_get_pixels(wlc_handle output) {
 	ipc_get_pixel_requests = unhandled;
 }
 
+static bool is_text_target(const char *target) {
+	return (strncmp(target, "text/", 5) == 0
+		|| strcmp(target, "UTF8_STRING") == 0
+		|| strcmp(target, "STRING") == 0
+		|| strcmp(target, "TEXT") == 0
+		|| strcmp(target, "COMPOUND_TEXT") == 0);
+}
+
+static void release_clipboard_request(struct get_clipboard_request *req) {
+	if (--(*req->pending) == 0) {
+		const char *str = json_object_to_json_string(req->json);
+		ipc_send_reply(req->client, str, (uint32_t)strlen(str));
+		json_object_put(req->json);
+	}
+
+	free(req->type);
+	free(req->buf);
+	wlc_event_source_remove(req->fd_event_source);
+	wlc_event_source_remove(req->timer_event_source);
+	close(req->fd);
+	free(req);
+}
+
+static int ipc_selection_data_cb(int fd, uint32_t mask, void *data) {
+	assert(data);
+	struct get_clipboard_request *req = (struct get_clipboard_request *)data;
+
+	if (mask & WLC_EVENT_ERROR) {
+		sway_log(L_ERROR, "Selection data fd error");
+		goto error;
+	}
+
+	if (mask & WLC_EVENT_READABLE) {
+		static const unsigned int max_size = 8192 * 1024;
+		int amt = 0;
+
+		do {
+			int size = req->buf_size - req->buf_position;
+			int amt = read(fd, req->buf + req->buf_position, size - 1);
+			if (amt < 0) {
+				if (errno == EAGAIN) {
+					return 0;
+				}
+
+				sway_log_errno(L_INFO, "Failed to read from clipboard data fd");
+				goto release;
+			}
+
+			req->buf_position += amt;
+			if (req->buf_position >= req->buf_size - 1) {
+				if (req->buf_size >= max_size) {
+					sway_log(L_ERROR, "get_clipbard: selection data too large");
+					goto error;
+				}
+				char *next = realloc(req->buf, req->buf_size *= 2);
+				if (!next) {
+					sway_log_errno(L_ERROR, "get_clipboard: realloc data buffer failed");
+					goto error;
+				}
+
+				req->buf = next;
+			}
+		} while(amt != 0);
+
+		req->buf[req->buf_position] = '\0';
+
+		json_object *obj = json_object_new_object();
+		json_object_object_add(obj, "success", json_object_new_boolean(true));
+		if (is_text_target(req->type)) {
+			json_object_object_add(obj, "content", json_object_new_string(req->buf));
+			json_object_object_add(req->json, req->type, obj);
+		} else {
+			size_t outlen;
+			char *b64 = b64_encode(req->buf, req->buf_position, &outlen);
+			json_object_object_add(obj, "content", json_object_new_string(b64));
+			free(b64);
+
+			char *type = malloc(strlen(req->type) + 8);
+			strcat(type, ";base64");
```

Comment:
yes that could be implemented that way (if it is really important enough). It would make the request and answer mimetypes at least match each other again, which was my primary point.